fiftyone/core/odm/database.py

```diff
@@ -1484,6 +1484,98 @@ def delete_runs(name, dry_run=False):
     )
 
 
+def get_indexed_values(
+    dataset, field_or_index, pipeline=None, values_only=False
+):
+    """Returns the values of the field for all samples in the given collection that are covered by the index.
+    Raises an error if the field is not indexed.
+
+    Args:
+        dataset: the dataset to query
+        field_or_index: the name of the indexed field to query or the index key for a compound index
+        pipeline: an optional pipeline to apply to the collection. For performance,
+            this should not include any fields not covered by the index.
+        values_only (False): whether to remove field names from the resulting list.
+            If True, the field names are removed and only the values will be returned as a list for each sample.
+            If False, the field names are preserved and the values will be returned as a dict for each sample.
+    Returns:
+        a list of values for the specified field or index keys for each sample sorted in the same order as the index
+    """
+    if field_or_index == "id":
+        if values_only:
+            return [
+                str(doc["_id"])
+                for doc in _iter_indexed_values(
+                    dataset, field_or_index, pipeline
+                )
+            ]
+        return [
+            {"id": str(doc["_id"])}
+            for doc in _iter_indexed_values(dataset, field_or_index, pipeline)
+        ]
+    if values_only:
+        # If values_only is True, we need to extract the values from the dict
+        if "_1_" not in field_or_index:
+            # Single field index
+            return [
+                doc[field_or_index]
+                for doc in _iter_indexed_values(
+                    dataset, field_or_index, pipeline
+                )
+            ]
+        return [
+            list(doc.values())
+            for doc in _iter_indexed_values(dataset, field_or_index, pipeline)
+        ]
+    return _iter_indexed_values(dataset, field_or_index, pipeline).to_list()
+
+
+def _iter_indexed_values(dataset, field_or_index, pipeline=None):
+    collection = dataset._sample_collection
+    if (
+        (field_or_index != "_id")
+        and (
+            field_or_index
+            not in (default_indexes := dataset._get_default_indexes())
+        )
+        and (
+            field_or_index not in (indexes := dataset.get_index_information())
```

Comment:
We might want to run a couple of scenarios with index creation for large indexes; if an index is under construction and we try to query it, MongoDB might make us wait for index creation to complete. We can use `get_index_information(include_stats=True)` to check if an index is in progress.

Reply:
The block/exclusive lock is only at the start of index creation and at the end (https://www.mongodb.com/docs/manual/core/index-creation/#index-build-process). Otherwise it's in the background and things are slow, but there's nothing we can do about that and it's no slower than calling values(). Also, during the long background index building, include_stats doesn't report it as in progress. To see that you have to query db.currentOp() and check for index creation.